pipeline_dp/contribution_bounders.py

```diff
@@ -0,0 +1,145 @@
+        """See docstrings for this class and for the base class bound_contributions."""
+        max_partitions_contributed = params.max_partitions_contributed
+        max_contributions_per_partition = params.max_contributions_per_partition
+        col = backend.map_tuple(
+            col, lambda pid, pk, v: ((pid, pk), v),
+            "Rekey to ( (privacy_id, partition_key), value))")
+        col = backend.sample_fixed_per_key(
+            col, params.max_contributions_per_partition,
+            "Sample per (privacy_id, partition_key)")
+        report_generator.add_stage(
+            f"Per-partition contribution bounding: randomly selected not "
+            f"more than {max_contributions_per_partition} contributions")
+        # ((privacy_id, partition_key), [value])
+        col = backend.map_values(
+            col, aggregate_fn,
+            "Apply aggregate_fn after per partition bounding")
+        # ((privacy_id, partition_key), accumulator)
+        # Cross partition bounding
+        col = backend.map_tuple(
+            col, lambda pid_pk, v: (pid_pk[0], (pid_pk[1], v)),
+            "Rekey to (privacy_id, (partition_key, accumulator))")
+        col = backend.sample_fixed_per_key(col,
+                                           params.max_partitions_contributed,
+                                           "Sample per privacy_id")
+
+        report_generator.add_stage(
+            f"Cross-partition contribution bounding: randomly selected not more"
+            f" than {max_partitions_contributed} partitions per privacy id")
+
+        # (privacy_id, [(partition_key, accumulator)])
+        def unnest_cross_partition_bound_sampled_per_key(pid_pk_v):
+            pid, pk_values = pid_pk_v
+            return (((pid, pk), v) for (pk, v) in pk_values)
+
+        return backend.flat_map(col,
+                                unnest_cross_partition_bound_sampled_per_key,
+                                "Unnest")
+
+
+class SamplingPerPrivacyIdContributionBounder(ContributionBounder):
+    """Bounds the total contributions of a privacy_id.
+
+    If a privacy_id contributes more than max_contributions, then
+    max_contributions contributions are uniformly sampled, otherwise all
+    contributions are kept.
+
+    Warning: This contribution bounder does group_by by privacy_id and processes
+    each privacy_id's data on one worker. In case if there are some privacy_ids
+    which contributes too much, it might lead to stale worker jobs.
+    """
+
+    def bound_contributions(self, col, params, backend, report_generator,
+                            aggregate_fn):
+        """See docstrings for this class and for the base class bound_contributions."""
+        max_contributions = params.max_contributions
+        col = backend.map_tuple(
+            col, lambda pid, pk, v: (pid, (pk, v)),
+            "Rekey to ((privacy_id), (partition_key, value))")
+        col = backend.sample_fixed_per_key(col, max_contributions,
```

Comment:
This wouldn't work for Sums if we want to do a per-partition/per-user sum before doing contribution bounding like we do in go, right?

Reply:
Good catch, correct, it wouldn't work for sum and `max_sum_per_partition`, not all ContributorBound are compatible with all Combiners. `max_sum_per_partition` it's not yet implement, it will require a separate ContributionBounder